### PR TITLE
Replace MOQ with NSubstitute

### PR DESCRIPTION
**Pull request by aydjay**
As per the huge discussion chain here:

https://github.com/moq/moq/issues/1372

Replace MOQ with NSubstitute

### CODE REVIEW

**Comment by tillig** [comment]
Thanks! I'm going to hold on this a moment and see how this pans out. The current version we're referencing doesn't have the privacy issues noted and, from a short term perspective, we can always just pin `<4.20.0` as the package reference. If this changes and a newer version of Moq comes out without this issue, we can "jump" to the new one, still saving some time. And if it doesn't... well, then this PR will come in handy.

File: test/Autofac.Test/Features/Decorators/DecoratorMiddlewareTests.cs [hunk 1]
```diff
@@ -14,19 +14,19 @@ public class DecoratorMiddlewareTests
...
+        contextMock.Instance.Returns(new object());
+        contextMock.Registration.Returns(registrationMock);
+        registrationMock.Options.Returns(RegistrationOptions.None);
 
-        middleware.Execute(contextMock.Object, context => { });
+        middleware.Execute(contextMock, context => { });
 
-        contextMock.Verify(context => context.Instance, Times.Once);
-        contextMock.Verify(context => context.Registration.Options, Times.Once);
+        _ = contextMock.Received(1).Instance;
```
Comment (tillig) [inline test/Autofac.Test/Features/Decorators/DecoratorMiddlewareTests.cs:26]: This discard construct reads kind of weird. Is there an error of some nature if the discard part is omitted?

Comment (aydjay) [reply test/Autofac.Test/Features/Decorators/DecoratorMiddlewareTests.cs:26]: Good spot! My mistake.

File: test/Autofac.Test/Core/Resolving/Middleware/CircularDependencyMiddlewareTests.cs [hunk 2]
```diff
@@ -11,12 +11,12 @@ public class CircularDependencyMiddlewareTests
...
-        var resolveRequestContextMock = new Mock<ResolveRequestContext>();
+        var resolveRequestContextMock = Substitute.For<ResolveRequestContext>();
         var middleware =
             new CircularDependencyDetectorMiddleware(CircularDependencyDetectorMiddleware.DefaultMaxResolveDepth);
 
-        middleware.Execute(resolveRequestContextMock.Object, context => { });
+        middleware.Execute(resolveRequestContextMock, context => { });
 
-        resolveRequestContextMock.Verify(context => context.Operation.RequestDepth, Times.Never);
+        _ = resolveRequestContextMock.Received().Operation.RequestDepth;
```
Comment (tillig) [inline test/Autofac.Test/Core/Resolving/Middleware/CircularDependencyMiddlewareTests.cs:20]: Looks like there's another of those discard items here.

Comment (aydjay) [reply test/Autofac.Test/Core/Resolving/Middleware/CircularDependencyMiddlewareTests.cs]: Corrected, as requested.

**Review by tillig — APPROVED** [review]
Thanks again, this is great!

**Comment by aydjay** [comment]
> Thanks again, this is great!

You're welcome, pleasure to work with you.

**Pull request merged**